### PR TITLE
[SPARK-32400][SQL]  Improve test coverage of HiveScriptTransformationExec

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?

1. Extract common test case (no serde) to BasicScriptTransformationExecSuite
2. Add more test case for no serde mode about supported data type and behavior in `BasicScriptTransformationExecSuite`
3. Add more test case for hive serde mode about supported type and behavior in `HiveScriptTransformationExecSuite`

### Why are the changes needed?
Improve test coverage of Script Transformation

### Does this PR introduce _any_ user-facing change?
NO

### How was this patch tested?
Added UT

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
FYI cc @maropu @cloud-fan

**Comment by maropu** [comment]
Thanks for the work, @AngersZhuuuu . Could you please fill the description as @HyukjinKwon suggested in the jira. I think that helps for better traceability.

**Comment by AngersZhuuuu** [comment]
> Thanks for the work, @AngersZhuuuu . Could you please fill the description as @HyukjinKwon suggested in the jira. I think that helps for better traceability.

Yea, updated.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala [hunk 1]
```diff
@@ -1063,6 +1063,9 @@ private[hive] trait HiveInspectors {
       case DateType => dateTypeInfo
       case TimestampType => timestampTypeInfo
       case NullType => voidTypeInfo
+      case dt =>
+        throw new AnalysisException(
+          s"${dt.getClass.getSimpleName.replace("$", "")} cannot be converted to Hive TypeInfo")
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala:1068]: nit: `dt.toString` should be good enough

Comment (AngersZhuuuu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala:1068]: > nit: `dt.toString` should be good enough

Updated, thanks

File: sql/core/src/test/scala/org/apache/spark/sql/execution/BaseScriptTransformationSuite.scala [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+            AttributeReference("d", StringType)(),
+            AttributeReference("e", new SimpleTupleUDT)()),
+          child = child,
+          ioschema = defaultIOSchema
+        ),
+        df.select('a, 'b.cast("string"), 'c.cast("string"), 'd.cast("string"), 'e).collect())
+    }
+  }
+
+  test("SPARK-32106: TRANSFORM should respect DATETIME_JAVA8API_ENABLED (no serde)") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/BaseScriptTransformationSuite.scala:291]: Plz update the jira ID.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala [hunk 3]
```diff
@@ -1063,6 +1063,9 @@ private[hive] trait HiveInspectors {
       case DateType => dateTypeInfo
       case TimestampType => timestampTypeInfo
       case NullType => voidTypeInfo
+      case dt =>
+        throw new AnalysisException(
+          s"${dt.toString} cannot be converted to Hive TypeInfo")
```
Comment (maropu) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala:1068]: `toString` -> `catalogString`

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationSuite.scala [hunk 4]
```diff
@@ -195,75 +169,146 @@ class HiveScriptTransformationSuite extends SparkPlanTest with SQLTestUtils with
...
+          ioschema = hiveIOSchema)
+        SparkPlanTest.executePlan(plan, hiveContext)
+      }.getMessage
+      assert(e2.contains("org.apache.spark.sql.types.TestUDT$MyDenseVectorUDT") &&
+        e2.contains("cannot be converted to Hive TypeInfo"))
     }
   }
 
-  override def output: Seq[Attribute] = child.output
+  test("SPARK-32106: TRANSFORM doesn't support" +
```
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationSuite.scala:283]: ditto: plz update the jira ID.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationSuite.scala [hunk 5]
```diff
@@ -195,75 +169,146 @@ class HiveScriptTransformationSuite extends SparkPlanTest with SQLTestUtils with
...
 
-  override def outputPartitioning: Partitioning = child.outputPartitioning
+      val e1 = intercept[SparkException] {
+        sql(
+          """
+            |SELECT TRANSFORM(a, b) USING 'cat' AS (a, b)
+            |FROM v
+          """.stripMargin).collect()
+      }.getMessage
+      assert(e1.contains(
```
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationSuite.scala:300]: nit: you don't need the line break here.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationSuite.scala [hunk 6]
```diff
@@ -195,75 +169,146 @@ class HiveScriptTransformationSuite extends SparkPlanTest with SQLTestUtils with
...
+          input = Seq(df.col("a").expr, df.col("b").expr),
+          script = "cat",
+          output = Seq(
+            AttributeReference("a", IntegerType)(),
+            AttributeReference("b", CalendarIntervalType)()),
+          child = df.queryExecution.sparkPlan,
+          ioschema = hiveIOSchema)
+        SparkPlanTest.executePlan(plan, hiveContext)
+      }.getMessage
+      assert(e1.contains(
```
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationSuite.scala:264]: nit: you don't need the line break here.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationSuite.scala [hunk 7]
```diff
@@ -195,75 +169,146 @@ class HiveScriptTransformationSuite extends SparkPlanTest with SQLTestUtils with
...
-      assert(TaskContext.get() != null) // Make sure that TaskContext is defined.
-      Thread.sleep(1000) // This sleep gives the external process time to start.
-      throw new IllegalArgumentException("intentional exception")
+  test("SPARK-32106: TRANSFORM doesn't support CalenderIntervalType/UserDefinedType (hive serde)") {
+    assume(TestUtils.testCommandAvailable("/bin/bash"))
+    withTempView("v") {
+      val df = Seq(
+        (1, new CalendarInterval(7, 1, 1000), new TestUDT.MyDenseVector(Array(1, 2, 3))),
+        (1, new CalendarInterval(7, 1, 1000), new TestUDT.MyDenseVector(Array(1, 2, 3))))
+        .toDF("a", "b", "c")
```
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationSuite.scala:250]: nit: How about following the same format? https://github.com/apache/spark/pull/29401/files#diff-a62183f534dac066bacc96d9e0d3885bR118-R122

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationSuite.scala [hunk 8]
```diff
@@ -195,75 +169,146 @@ class HiveScriptTransformationSuite extends SparkPlanTest with SQLTestUtils with
...
 
-  override def output: Seq[Attribute] = child.output
+  test("SPARK-32106: TRANSFORM doesn't support" +
+    " CalenderIntervalType/UserDefinedType end to end (hive serde)") {
+    assume(TestUtils.testCommandAvailable("/bin/bash"))
+    withTempView("v") {
+      val df = Seq(
+        (1, new CalendarInterval(7, 1, 1000), new TestUDT.MyDenseVector(Array(1, 2, 3))),
+        (1, new CalendarInterval(7, 1, 1000), new TestUDT.MyDenseVector(Array(1, 2, 3))))
+        .toDF("a", "b", "c")
```
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationSuite.scala:290]: nit: How about following the same format? https://github.com/apache/spark/pull/29401/files#diff-a62183f534dac066bacc96d9e0d3885bR118-R122

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationSuite.scala (hunk 6 above)
Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationSuite.scala:264]: Done

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationSuite.scala (hunk 5 above)
Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationSuite.scala:300]: Done

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationSuite.scala (hunk 4 above)
Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationSuite.scala:283]: Done

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala (hunk 3 above)
Comment (AngersZhuuuu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala:1068]: Done

File: sql/core/src/test/scala/org/apache/spark/sql/execution/BaseScriptTransformationSuite.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/BaseScriptTransformationSuite.scala:291]: Done

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationSuite.scala (hunk 7 above)
Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationSuite.scala:250]: Done

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationSuite.scala (hunk 8 above)
Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationSuite.scala:290]: Done

**Comment by maropu** [comment]
Looks okay and I have no more comment.

**Review by maropu — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**